src/rt/aaA.d

```diff
@@ -474,6 +474,23 @@ extern (C) void _aaClear(AA aa) pure nothrow
     }
 }
 
+/// Reserve AA
+extern (C) void _aaReserve(AA* aa, const TypeInfo_AssociativeArray ti, size_t ndim)
+{
+    ndim = nextpow2(ndim);
```

Comment:
The buckets size used should be large enough that if ndim elements are added, the bucket array will not be grown further (ndim function arg value). However, the choices here don't achieve that. Specifically, the bucket array will be grown if it becomes 80% full (GROW_NUM/GROW_DEN). nextpow2(ndim) will not always need this requirement. It will sometimes be necessary to move one power of two further to achieve this. e.g. If the arg is 500, the bucket size selected should be 1024, not 512. I suggest testing for this after the `ndim = nextpow2(ndim)` line, if it doesn't meet it, find the next power of two. That should handle the first two cases (`aa.impl is null`, `ndim <= aa.dim`). For the third case tested `(aa.used * GROW_DEN > ndim * GROW_NUM)`, it is less clear to me when that would get triggered.

Reply:
I think applying `ndim = (ndim * GROW_DEN + GROW_NUM - 1) / GROW_NUM;` before finding the next power of 2 should be good enough. 
I'm not 100% sure, but you might also have to take the deleted entries into account when comparing with aa.dim or aa.used.

Reply:
One thing to be cautious of is that `reserve` should never remove buckets. Due to the logic below, it won't, but if the logic is moved around, it could easily happen.

Reply:
Regarding the power-of-two allocation sizes: Best I can tell all the existing bucket allocation code paths use `nextpow2` to set the bucket size when allocating. Not clear if that is important. If it is, the

    ndim = (aa.used * GROW_DEN + GROW_NUM - 1) / GROW_NUM;

line is not following it. However, it the power-of-two component is not important, there may be an opportunity with `reserve`. Assuming that a call to `reserve` is a hint about expected element count, it might be better to choose the size based on fill ratio rather than power-of-two. For example, reserving 7 million entries will pick a power-of-two of 16777216, a 42% fill ratio. A 70% fill ratio of 10 million would save space and improve locality of reference.